docs: lsp, fix errors, gen_help_html features

justinmk opened the pull request:
- merge lsp-handler documentation
- fix various invalid :help syntax errors in preparation for https://github.com/neovim/tree-sitter-vimdoc/pull/97

justinmk committed:
docs: lsp, vim_diff

- quickstart
- mark lsp.txt as `new_layout`
- remove lsp-handler documentation for notifications: they don't have
  handlers because they don't have server responses.

justinmk commented on runtime/doc/lsp.txt:
```diff
@@ -243,59 +239,24 @@ For |lsp-response|, each |lsp-handler| has this signature: >
-
-        You can use |vim.lsp.rpc.rpc_response_error()| to create this object.
-
-For |lsp-notification|, each |lsp-handler| has this signature: >
```
Notification handler is very similar to the request case, so it's probably more helpful to combine them than split them.

justinmk replied:
> The method also doesn't take a handler:

Yeah I was wondering about that :) So the doc should just be removed, right?

What is implied by these tests?: https://github.com/neovim/neovim/blob/e27377e33e029a9c2efb282d4c4e51effd76dd88/test/functional/plugin/lsp_spec.lua#L281

That's just testing for the message, not a handler of any sort?

mfussenegger replied:
> So the doc should just be removed, right?

Yes

> That's just testing for the message, not a handler of any sort?

It's only testing that the notification sent by the client arrives at the server.

justinmk committed:
fix(docs): vimdoc syntax errors

Since https://github.com/neovim/tree-sitter-vimdoc/pull/97
the many cases of *.foo cause parser errors. But even before that, these
were erroneously highlighted as (argument), so fixing them is good.

justinmk committed:
feat(gen_help_html): ignore pi_netrw.txt errors

IDGAF about netrw

justinmk committed:
feat(gen_help_html): non-default vimdoc.so parser

Callers can specify a non-default vimdoc.so file path.

justinmk commented on runtime/doc/autocmd.txt:
```diff
@@ -1417,24 +1417,24 @@ When there is a matching "*Cmd" autocommand, it is assumed it will do the
-Note that the *WritePost commands should undo any changes to the buffer that
-were caused by the *WritePre commands; otherwise, writing the file will have
+Note that the "*WritePost" commands should undo any changes to the buffer that
+were caused by the "*WritePre" commands; otherwise, writing the file will have
```
Even before https://github.com/neovim/tree-sitter-vimdoc/pull/97 these were not rendered correctly: they were treated as `(argument)` which conceals the "*".

justinmk commented on runtime/doc/lsp.txt:
```diff
@@ -173,129 +162,72 @@ when creating a new client. Keys are LSP method names: >vim
+Nvim LSP client (but depends on server support):
+
+- callHierarchy/incomingCalls
+- callHierarchy/outgoingCalls
```
Now we can generate `runtime/doc/lsp.txt` as a "flow" layout instead of fixed-width.

clason commented on runtime/doc/eval.txt:
```diff
@@ -39,7 +39,7 @@ List		An ordered sequence of items, see |List| for details.
 Dictionary	An associative, unordered array: Each entry has a key and a
 		value. |Dictionary|
-		Examples:
+		Examples: >
```
Might want to get into the habit of always using explicit closing markers (`<`) instead of upstream's Hail Mary, at least for those markers we add ourselves.

(One day we will go through all the docs and sanitize them, but today is not that day.)

justinmk replied:
Really? I like the "feature" of omitting the closing `<` because it lets the `}` motion work more often.

clason replied:
Ah, I was not aware of that. Is it just the lack of empty line? (We can add one after the closing tag, of course, for extra cake.)

I've been religious about adding them to Nvim docs I touch, on principle as well as because sometimes that fixes (HTML) rendering issues.

justinmk replied:
> Is it just the lack of empty line? (We can add one after the closing tag, of course, for extra cake.)

that's part of it, but also just the closing `<` is unnecessary most of the time so it's noise.

otoh the `}` motion should probably handle "conceal" intellligently.

> I've been religious about adding them to Nvim docs I touch, on principle as well as because sometimes that fixes (HTML) rendering issues.

parser should handle it except in the middle of a list 😔

clason commented on runtime/doc/filetype.txt:
```diff
@@ -881,7 +881,7 @@ file: |ft_sql.txt|.
 TEX						*ft-tex-plugin* *g:tex_flavor*
 
-If the first line of a *.tex file has the form >
+If the first line of a "*.tex" file has the form >
```
`*.tex`? (not sure how much you care about consistency here)

justinmk replied:
Yeah this was intentional, I think quotes are what is intended for most of these cases, except when in a "table". Quotes also don't avoid parser errors if there is whitespace within the quotes.

clason replied:
Sorry, markdown got interpreted, which made the comment... idiotic.

I meant everywhere else in this commit, you've used backticks around *.foo, while (only) here you use double quotes.

justinmk replied:
fixed

clason commented on runtime/doc/syntax.txt:
```diff
@@ -1427,13 +1427,13 @@ Euphoria version 3.1.1 (https://www.rapideuphoria.com/) is still necessary
 (https://www.openeuphoria.org/) does not support.
 
-The following file extensions are auto-detected as Euphoria file type:
+The following file extensions are auto-detected as Euphoria file type: >
```
same re: closing markers

clason commented on runtime/doc/syntax.txt:
```diff
@@ -1469,11 +1469,11 @@ ELIXIR						*elixir.vim* *ft-elixir-syntax*
 maintainable applications.
 
-The following file extensions are auto-detected as Elixir file types:
+The following file extensions are auto-detected as Elixir file types: >
```
and here

clason commented on runtime/doc/syntax.txt:
```diff
@@ -4473,19 +4473,19 @@ Notes:
-Example (match a comment but don't highlight the /* and */): >
+Example (match a comment but don't highlight the `/* and */`): >
   :syntax region Comment start="/\*"hs=e+1 end="\*/"he=s-1
-<
```
actually, I'd suggest keeping this as two separate code blocks, so we can (in the future) mark the first as `vim`

clason commented on runtime/doc/syntax.txt:
```diff
@@ -5481,14 +5481,14 @@ memory Vim will consume.
 must use Universal Ctags (https://ctags.io) or Exuberant ctags.
 
-Put these lines in your Makefile:
+Put these lines in your Makefile: >
```
closing marker -- and arguably, _if_ we add new markers to Vim help files, we might as well annotate them (although _here_ it doesn't make much sense until we ship a makefile parser)
